### ISSUE TITLE
Cannot install KingPhisher onto CentOS7

### ISSUE THREAD

**Issue by dookgit** [securestate/king-phisher#353]
## Issue Description
I've been trying to install KingPhisher onto a CentOS7 server, and it keeps failing on the "Synchronizing King Phisher's pipenv environment"

It delays on that step for a while, then returns with:

pipenv encountered the following error:
Installing dependencies from Pipfile.lock (49fa41)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment

It then continues on, and tries to start KingPhisher service, but fails.

## Reproduction Steps
Deploy basic CentOS7 server
Download install.sh and run (as root)

## Environment Details

CentOS Linux release 7.6.1810 (Core)
King Phisher Version: Latest, as of today

## Error Details / Stack Trace
INFO: Synchronizing King Phisher's pipenv environment
pipenv encountered the following error:
Installing dependencies from Pipfile.lock (49fa41)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: centos.mirror.serversaustralia.com.au
 * epel: d2lzkl7pfhq30w.cloudfront.net
 * extras: centos.mirror.serversaustralia.com.au
 * ius: syd.mirror.rackspace.com
 * updates: mirror.ventraip.net.au
Package postgresql-server-9.2.24-1.el7_5.x86_64 already installed and latest version
Nothing to do
Data directory is not empty!

INFO: Creating King Phisher admin group: 'king-phisher'
INFO: Configuring the PostgreSQL server
INFO: PostgreSQL configuration file found at /var/lib/pgsql/data/pg_hba.conf
INFO: Created the PostgreSQL user 'king_phisher' with a random password
INFO: Installing the King Phisher systemd service file in /lib/systemd/system/
INFO: Starting the King Phisher service
Created symlink from /etc/systemd/system/multi-user.target.wants/king-phisher.service to /usr/lib/systemd/system/king-phisher.service.
Job for king-phisher.service failed because a timeout was exceeded. See "systemctl status king-phisher.service" and "journalctl -xe" for details.

When running KingPhisher manually (/usr/bin/python3 /opt/king-phisher/KingPhisherServer /opt/king-phisher/server_config.yml)

Loading .env environment variables…
Traceback (most recent call last):
  File "/usr/lib64/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/king-phisher/king_phisher/server/__main__.py", line 42, in <module>
    from king_phisher import color
  File "/opt/king-phisher/king_phisher/color.py", line 41, in <module>
    import termcolor
ImportError: No module named 'termcolor'

**Comment by wolfthefallen** [comment]
King Phisher now utilizes `pipenv` to maintain packages and prevent issues with system level packages. You will need to change directories to King Phisher directory in this case `/opt/king-phisher` then run the server from there, please run in `DEBUG` mode and provide us the output to help you trouble shoot the issue.

Commands to change to King Phisher directory and run the server in debug mode.
```
cd /opt/king-phisher
sudo ./KingPhisherServer -L DEBUG -f ./server_config.yml
```

**Comment by dookgit** [comment]
I am having issues starting KingPhisher, but i'm fairly sure it's related to it not installing correctly, and the pipenv.

I've manually created the /opt/king-phisher directory, run install.sh from there, but same issue.
I also tried cloning the repo into /opt/king-phisher and running install.sh from there too, and again, fails on "INFO: Synchronizing King Phisher's pipenv environment"  with:

pipenv encountered the following error:
Installing dependencies from Pipfile.lock (49fa41)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment


When running KingPhisher server with the DEBUG switch i get:

DEBUG    target diretory: /opt/king-phisher
INFO     checking for the pipenv environment
DEBUG    pipenv path: '/usr/bin/pipenv'
DEBUG    pipenv Pipfile: /opt/king-phisher/Pipfile
Loading .env environment variables…
Traceback (most recent call last):
  File "/usr/lib64/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/king-phisher/king_phisher/server/__main__.py", line 42, in <module>
    from king_phisher import color
  File "/opt/king-phisher/king_phisher/color.py", line 41, in <module>
    import termcolor
ImportError: No module named 'termcolor'



Thanks

**Comment by dookgit** [comment]
Just a quick update, i'm a noob with anything Python related, but i tried running "pipenv install" from the /opt/king-phisher directory and got the following:

Installing dependencies from Pipfile.lock (49fa41)…
An error occurred while installing pycairo==1.18.0 --hash=sha256:abd42a4c9c2069febb4c38fe74bfc4b4a9d3a89fea3bc2e4ba7baff7a20f783f! Will try again.
An error occurred while installing https://github.com/matplotlib/basemap/archive/v1.2.0rel.tar.gz#egg=basemap! Will try again.
An error occurred while installing python-dateutil==2.7.2 --hash=sha256:3220490fb9741e2342e1cf29a503394fdac874bc39568288717ee67047ff29df --hash=sha256:9d8074be4c993fbe4947878ce593052f71dac82932a677d49194d8ce9778002e! Will try again.
An error occurred while installing pynacl==1.3.0 --hash=sha256:05c26f93964373fc0abe332676cb6735f0ecad27711035b9472751faa8521255 --hash=sha256:0c6100edd16fefd1557da078c7a31e7b7d7a52ce39fdca2bec29d4f7b6e7600c --hash=sha256:0d0a8171a68edf51add1e73d2159c4bc19fc0718e79dec51166e940856c2f28e --hash=sha256:1c780712b206317a746ace34c209b8c29dbfd841dfbc02aa27f2084dd3db77ae --hash=sha256:2424c8b9f41aa65bbdbd7a64e73a7450ebb4aa9ddedc6a081e7afcc4c97f7621 --hash=sha256:2d23c04e8d709444220557ae48ed01f3f1086439f12dbf11976e849a4926db56 --hash=sha256:30f36a9c70450c7878053fa1344aca0145fd47d845270b43a7ee9192a051bf39 --hash=sha256:37aa336a317209f1bb099ad177fef0da45be36a2aa664507c5d72015f956c310 --hash=sha256:4943decfc5b905748f0756fdd99d4f9498d7064815c4cf3643820c9028b711d1 --hash=sha256:57ef38a65056e7800859e5ba9e6091053cd06e1038983016effaffe0efcd594a --hash=sha256:5bd61e9b44c543016ce1f6aef48606280e45f892a928ca7068fba30021e9b786 --hash=sha256:6482d3017a0c0327a49dddc8bd1074cc730d45db2ccb09c3bac1f8f32d1eb61b --hash=sha256:7d3ce02c0784b7cbcc771a2da6ea51f87e8716004512493a2b69016326301c3b --hash=sha256:a14e499c0f5955dcc3991f785f3f8e2130ed504fa3a7f44009ff458ad6bdd17f --hash=sha256:a39f54ccbcd2757d1d63b0ec00a00980c0b382c62865b61a505163943624ab20 --hash=sha256:aabb0c5232910a20eec8563503c153a8e78bbf5459490c49ab31f6adf3f3a415 --hash=sha256:bd4ecb473a96ad0f90c20acba4f0bf0df91a4e03a1f4dd6a4bdc9ca75aa3a715 --hash=sha256:e2da3c13307eac601f3de04887624939aca8ee3c9488a0bb0eca4fb9401fc6b1 --hash=sha256:f67814c38162f4deb31f68d590771a29d5ae3b1bd64b75cf232308e5c74777e0! Will try again.
An error occurred while installing tzlocal==1.5.1 --hash=sha256:4ebeb848845ac898da6519b9b31879cf13b6626f7184c496037b818e238f2c4e! Will try again.
An error occurred while installing graphene-sqlalchemy==2.1.0 --hash=sha256:5bfc93986c5e6001861e8cdfe33eeb5094ca1f30efde6822387e13070ec3ab82 --hash=sha256:8260a128f360eda5437147a3cb72809e267fa8e2464f9996ff95ddc47239dcec! Will try again.
An error occurred while installing promise==2.2.1 --hash=sha256:2ebbfc10b7abf6354403ed785fe4f04b9dfd421eb1a474ac8d187022228332af --hash=sha256:348f5f6c3edd4fd47c9cd65aed03ac1b31136d375aa63871a57d3e444c85655c! Will try again.
An error occurred while installing geoip2==2.9.0 --hash=sha256:a37ddac2d200ffb97c736da8b8ba9d5d8dc47da6ec0f162a461b681ecac53a14 --hash=sha256:f7ffe9d258e71a42cf622ce6350d976de1d0312b9f2fbce3975c7d838b57ecf0! Will try again.
An error occurred while installing cycler==0.10.0 --hash=sha256:1d8a5ae1ff6c5cf9b93e8811e581232ad8920aeec647c37316ceac982b08cb2d --hash=sha256:cd7b2d1018258d7247a71425e9f26463dfb444d411c39569972f4ce586b0c9d8! Will try again.
An error occurred while installing graphene==2.1.3 --hash=sha256:b8ec446d17fa68721636eaad3d6adc1a378cb6323e219814c8f98c9928fc9642 --hash=sha256:faa26573b598b22ffd274e2fd7a4c52efa405dcca96e01a62239482246248aa3! Will try again.
An error occurred while installing paramiko==2.4.2 --hash=sha256:3c16b2bfb4c0d810b24c40155dbfd113c0521e7e6ee593d704e84b4c658a1f3b --hash=sha256:a8975a7df3560c9f1e2b43dc54ebd40fd00a7017392ca5445ce7df409f900fcb! Will try again.
An error occurred while installing alembic==1.0.2 --hash=sha256:04bcb970ca8659c3607ddd8ffd86cc9d6a99661c9bc590955e8813c66bfa582b! Will try again.
An error occurred while installing graphql-relay==0.4.5 --hash=sha256:2716b7245d97091af21abf096fabafac576905096d21ba7118fba722596f65db! Will try again.
An error occurred while installing pygobject==3.28.3 --hash=sha256:250fb669b6ac64eba034cc4404fcbcc993717b1f77c29dff29f8c9202da20d55! Will try again.
An error occurred while installing graphql-core==2.1 --hash=sha256:889e869be5574d02af77baf1f30b5db9ca2959f1c9f5be7b2863ead5a3ec6181 --hash=sha256:9462e22e32c7f03b667373ec0a84d95fba10e8ce2ead08f29fbddc63b671b0c1! Will try again.
An error occurred while installing icalendar==4.0.3 --hash=sha256:07c2447a1d44cbb27c90b8c6a5c98e890cc1853c6223e2a52195cddec26c6356 --hash=sha256:83f7248b7485ddd29c7d69b706b21c441e34855d9c1d888939fd24aefdd9d19b! Will try again.
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 66/66 — 00:01:56
Installing initially failed dependencies…
  ☤  ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 16/16 — 00:00:58
[pipenv.exceptions.InstallError]:   File "/usr/lib/python3.5/site-packages/pipenv/cli/command.py", line 254, in install
[pipenv.exceptions.InstallError]:       editable_packages=state.installstate.editables,
[pipenv.exceptions.InstallError]:   File "/usr/lib/python3.5/site-packages/pipenv/core.py", line 1874, in do_install
[pipenv.exceptions.InstallError]:       keep_outdated=keep_outdated
[pipenv.exceptions.InstallError]:   File "/usr/lib/python3.5/site-packages/pipenv/core.py", line 1253, in do_init
[pipenv.exceptions.InstallError]:       pypi_mirror=pypi_mirror,
[pipenv.exceptions.InstallError]:   File "/usr/lib/python3.5/site-packages/pipenv/core.py", line 862, in do_install_dependencies
[pipenv.exceptions.InstallError]:       _cleanup_procs(procs, False, failed_deps_queue, retry=False)
[pipenv.exceptions.InstallError]:   File "/usr/lib/python3.5/site-packages/pipenv/core.py", line 681, in _cleanup_procs
[pipenv.exceptions.InstallError]:       raise exceptions.InstallError(c.dep.name, extra=err_lines)


So, is it failing to find/install some of the dependencies?  Cheers

**Comment by dookgit** [comment]
Just for my own sanity, i just spun up a Ubuntu 18.10 box, and was able to install KingPhisher fine, just by running ./install.sh.

So CentOS specific it seems.

**Comment by True-Demon** [comment]
I ran into similar issues with this while trying to install on a Debian server, but starting from scratch fixed the problem for me. As far as I could tell, it was because of something wrong with my environment, specifically pip or possibly python itself, which is strange but not impossible. Just to clarify, I've installed king-phisher on Fedora, Ubuntu, Debian, Kali, and Arch without issues, but once in a blue moon, it seems like something with Pip goes haywire and throws a wrench in the gears. Sometimes it's because of permission issues, sometimes it's a complete mystery.  ¯\_(ツ)_/¯

Start by blowing away the /opt/king-phisher directory completely, and install the essentials from yum (or DNF is that's what CentOS is using these days). 

`yum install python3 python3-dev python3-pip`
Verify your Python version by doing `python3 --version`
My current version is Python 3.7.2 but latest stable of Python 3.4.x should be fine. Whatever the case, just make sure you're on the latest stable version.
Verify your pip version by doing `pip3 -V`
Latest stable is 19.0.1 at the time of this issue.

At this point, I recommend you sudo to root and perform everything as the root user from here forward. This will make pip install everything globally and should make everything available for all users.

```
git clone https://github.com/securestate/king-phisher /opt/king-phisher
cd /opt/king-phisher/tools
chmod +x ./install.sh
./install.sh
```

If you still get errors related to the hashes or install process, print the output of `pipenv install --verbose` from /opt/king-phisher. Again, make sure you're installing as root.

**Comment by dookgit** [comment]
Hey, thanks for your help.

So python3 isn't installed in CentOS7 by default, so i did the following:

```
yum install -y https://centos7.iuscommunity.org/ius-release.rpm
yum install gcc openssl-devel bzip2-devel
yum install epel-release
yum install python36
python3.6 -m ensurepip
ln -s /usr/bin/python3.6 /usr/bin/python3
ln -s /usr/local/bin/pip3 /usr/bin/pip3

[root@template_kp_v1 ~]# python3 -V
Python 3.6.6
[root@template_kp_v1 ~]# pip3 -V
pip 10.0.1 from /usr/local/lib/python3.6/site-packages/pip (python 3.6)
```

I then ran the installer exactly as you suggested (as root) and it failed again on the pipenv.  When i ran 

`./KingPhisher --env-install -L DEBUG`

I got

`DEBUG    target diretory: /opt/king-phisher
INFO     checking for the pipenv environment
DEBUG    pipenv path: '/usr/local/bin/pipenv'
INFO     installing the pipenv environment
pipenv encountered the following error:
Installing dependencies from Pipfile.lock (49fa41)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment
INFO     removing the incomplete .venv directory`

I then ran pipenv install --verbose from /opt/king-phisher and got the attached (lot of output.)
[pipenv_install.log](https://github.com/securestate/king-phisher/files/2877733/pipenv_install.log)

I then reverted my VM, and stepped through the install.sh for CentOS7, installing the same version of python (3.5) and all the dependencies, and again it fails when running ./KingPhisher --env-install

Cheers

**Comment by wolfthefallen** [comment]
thank you for the information, it looks like a package is missing for basemaps that needs to be installed through the package manager. I'll need to kick it around some more to narrow down the missing packages

**Comment by dookgit** [comment]
Thanks, happy todo more testing my end too if it helps.

**Comment by True-Demon** [comment]
Looks like libgeos, except the naming convention doesn't match in yum/dnf as it does in apt like the error suggests. 
`yum/dnf install geos geos-devel` appears to resolve the dependency issue on Fedora. Should be the same for CentOS. Might be a necessary pre-requisite for RPM distros if it's not included in the basemap tarball that's pulled with the install.sh script.

**Comment by dookgit** [comment]
Hi, cheers for that.  I did a clean build again, and the  manually installed geos and geos-devel by using:
`yum install geos geos-devel`

Now when running:
`pipenv install --verbose`

It seems to go further, but failing on cairo
` .......
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.6
    creating build/lib.linux-x86_64-3.6/cairo
    copying cairo/__init__.py -> build/lib.linux-x86_64-3.6/cairo
    copying cairo/__init__.pyi -> build/lib.linux-x86_64-3.6/cairo
    copying cairo/py.typed -> build/lib.linux-x86_64-3.6/cairo
    warning: build_py: byte-compiling is disabled, skipping.

    running build_ext
    Package cairo was not found in the pkg-config search path.
    Perhaps you should add the directory containing `cairo.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'cairo' found
    Command '['pkg-config', '--print-errors', '--exists', 'cairo >= 1.13.1']' returned non-zero exit status 1.
    Running setup.py install for pycairo: finished with status 'error'
Cleaning up...
  Removing source in /tmp/pip-install-d5md_5m6/pycairo
Removed build tracker '/tmp/pip-req-tracker-zje42ax3'
Exception information:
Traceback (most recent call last):
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/pip/_internal/cli/base_command.py", line 179, in main
    status = self.run(options, args)
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/pip/_internal/commands/install.py", line 393, in run
    use_user_site=options.use_user_site,
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/pip/_internal/req/__init__.py", line 57, in install_given_reqs
    **kwargs
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/pip/_internal/req/req_install.py", line 945, in install
    spinner=spinner,
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/pip/_internal/utils/misc.py", line 761, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip._internal.exceptions.InstallationError: Command "/opt/king-phisher/.venv/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-d5md_5m6/pycairo/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-6epk7ozi/install-record.txt --single-version-externally-managed --compile --install-headers /opt/king-phisher/.venv/include/site/python3.6/pycairo" failed with error code 1 in /tmp/pip-install-d5md_5m6/pycairo/
Traceback (most recent call last):
  File "/usr/local/bin/pipenv", line 10, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/cli/command.py", line 254, in install
    editable_packages=state.installstate.editables,
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 1874, in do_install
    keep_outdated=keep_outdated
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 1253, in do_init
    pypi_mirror=pypi_mirror,
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 862, in do_install_dependencies
    _cleanup_procs(procs, False, failed_deps_queue, retry=False)
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 681, in _cleanup_procs
    raise exceptions.InstallError(c.dep.name, extra=err_lines)
pipenv.exceptions.InstallError: ERROR: ERROR: Package installation failed...`

**Comment by True-Demon** [comment]
Odd that it would fail on that dependency... 

You should try to rerun the king-phisher `tools/install.sh` script, since that should resolve any remaining dependencies, and your earlier problem of the basemap dependency is solved.

For good measure / curiosity's sake, can you run
`pip freeze | grep -i cairo` to verify whether it's installed and what version it's at.

Assuming it's not installed or behind it's version, you can try installing it manually.

Global Install: `sudo pip install pycairo`
User-only Install: `pip install pycairo --user`
Upgrade: `sudo pip install -U pycairo`

Either way, if you absolutely need to, you can always install each dependency that is mentioned in the **Piplock** file after a failed installation attempt. Pipenv will always default to using the locally installed version of whatever dependencies are needed, assuming they aren't pinned to a specific version and the local version exceeds the minimum required version. 

Let us know how it progresses.

**Comment by dookgit** [comment]
Hi folks.

OK, i've tried a few things, including install a few dependencies manually
`yum install -y cairo cairo-devel geos geos-devel`

And have managed to get to state where
`./KingPhisher --env-install -L DEBUG`
completes without any errors and also
`pipenv install --verbose`
completes successfully too, but when i try running ./install.sh it fails again on :
`INFO: Synchronizing King Phisher's pipenv environment
pipenv encountered the following error:
Installing dependencies from Pipfile.lock (49fa41)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment`

And if i try starting the service manually, it seems to fail on various options.
I might have to just give up installing on CentOS, as it looks like it's just gonna work.
Cheers

**Comment by wolfthefallen** [comment]
This issue has been resolved in commit 20ca83a25fa0c40423c60bdb60be508e8a7308be . This is currently in the dev branch. If you would like to get into beta testing King Phisher the instructions are located at [https://github.com/securestate/king-phisher/wiki/Updating-King-Phisher#beta-testing](https://github.com/securestate/king-phisher/wiki/Updating-King-Phisher#beta-testing) .

This issue was due to `cairo-devel`, `gobject-instrospection-devel`, `cairo-gobject` and `cairo-gobject-devel` not being installed for required packages pipenv was installing.
